File: solutions/typescript/resistor-color-trio/1/resistor-color-trio.ts
export function decodedResistorValue(colors: string[]): string {
  let mantissa = colors
    .slice(0, 2)
    .reverse()
    .reduce(
      (acc: number, color: string, index: number) => 
        acc + COLORS.indexOf(color) * 10 ** index,
      0
    )
  let order = COLORS.indexOf(colors[2]);
  return express(mantissa * 10 ** order);
}

function express(value: number): string {
  const prefix = [
    "",
    "kilo",
    "mega",
    "giga",
  ];
  let order = Math.floor(log1000(value));
  let scaledValue = value / (1000 ** order);
  let scaledUnits = `${prefix[order]}ohms`
  return `${scaledValue} ${scaledUnits}`;
}

function log1000(value: number): number {
  if (value == 0) return 0;
  return Math.log(value) / Math.log(1000);
}

const COLORS = [
  "black",
  "brown",
  "red",
  "orange",
  "yellow",
  "green",
  "blue",
  "violet",
  "grey",
  "white",
];

File: solutions/python/resistor-color-trio/1/resistor_color_trio.py
import math

COLOR_BY_VALUE = [
    "black",
    "brown",
    "red",
    "orange",
    "yellow",
    "green",
    "blue",
    "violet",
    "grey",
    "white",
]

PREFIX_FOR_THOUSANDS = [
    "",
    "kilo",
    "mega",
    "giga",
]

def value(colors):
    if len(colors) == 1:
        return COLOR_BY_VALUE.index(colors[0])
    return COLOR_BY_VALUE.index(colors[0]) * 10 + COLOR_BY_VALUE.index(colors[1])


def format(ohms):
    if ohms > 0:
        thousands = int(math.log10(ohms) / 3)
    else:
        thousands = 0
    if thousands > 0:
        base = ohms // (1000 ** thousands)
    else:
        base = ohms
    return str(base) + " " + PREFIX_FOR_THOUSANDS[thousands] + "ohms"


def label(colors):
    base = value(colors[:2])
    zeroes = value(colors[2:3])
    ohms = base * 10 ** zeroes
    return format(ohms)
